### PR TITLE
Standalone updater. Initial commit

### PR DESCRIPTION
**Pull request by VicDeo**
As per https://github.com/owncloud/core/issues/18721

Steps are taken from https://github.com/owncloud/core/issues/18721#issuecomment-143215159
- [ ]  upgrade:info                       Info. Show what is going to happen with explanation.
- [ ]  upgrade:checkSystem                System check. System health and if dependencies are OK (we also count the number of files and DB entries and give time estimations based on hardcoded estimation)
- [x]  upgrade:checkpoint                 Create or restore owncloud files
- [ ] upgrade:cleanCache                 clean all caches
- [x]  upgrade:detect                     Detect
  - 1. [x]    currently existing code, 
  - 2.  [x] version in config.php, 
  - 3.  [x] online available verison.  
  - [x] ASK what to do? (download, upgrade, abort, …)

- [x]  upgrade:disableNotShippedApps      Disable not shipped apps
- [x]  upgrade:enableNotShippedApps       try reenable and upgrade 3rdparty/non shipped apps. (one app after another)
- [x]  upgrade:executeCoreUpgradeScripts  execute core upgrade scripts [danger, might take long]   
- [x]  upgrade:maintenanceMode            Toggle maintenance mode
[danger, might take long]
- [ ]  upgrade:restartWebServer           Please Restart Web Server
- [ ]  upgrade:upgradeShippedApps         upgrade shipped apps [danger, might take long]

- [x]  upgrade:start      executes
 'upgrade:info'
'upgrade:checkSystem'
'upgrade:maintenanceMode --on' 
'upgrade:detect'
'upgrade:backupDb'
'upgrade:backupData'
'upgrade:preUpgradeRepair'
'upgrade:dbUpgrade simulation'],
'upgrade:dbUpgrade'
'upgrade:disableNotShippedApps'
'upgrade:executeCoreUpgradeScript'
'upgrade:upgradeShippedApps'
'upgrade:enableNotShippedApps'
'upgrade:cleanCache'
'upgrade:postUpgradeRepair'
'upgrade:restartWebServer'
'upgrade:updateConfig'
'upgrade:maintenanceMode --off'
'upgrade:postUpgradeCleanup'        

Not listed in the comment above but done:
- [x] downloading 
- [x] checking md5 for package
- [x] unpacking (zip only)
- [x] logging
- [x] check if there are enough permissions to replace files

Postponed for 9.0
moved to https://github.com/owncloud/updater/issues/204

### CODE REVIEW

**Comment by VicDeo** [comment]
@DeepDiver1975  @PVince81

**Comment by VicDeo** [comment]
As of today it is able to 
 - search for occ and version.php one level above it's path 
 - read it's config via occ config:list --output "json"
 - construct a feed URL from version,edition,build,etc 
 - download a file from the feed
 - get md5 checksum (by adding .md5 postfix to the package URL)
 - validate md5
 - location and file name is always the same so if there is already a package, it just validates md5
 - extract it via console unzip and fallback to PHP built-in ZipArchive (currently extraction path is /tmp but this is going to be changed very soon)

all exceptions are logged into updater.log with a full stacktrace
usage is limited by 8.2+ due to lack of implementation of occ config:list for versions below this one

File: nbproject/project.xml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (PVince81) [inline nbproject/project.xml]: remove "nbproject" folder ?

Comment (VicDeo) [inline nbproject/project.xml]: and add `.idea` ? ;)

**Comment by PVince81** [comment]
I scrolled through the code and it looks good so far.

Would it maybe be better to create a new app/repo for this ?
If we merge this PR we "destroy" the old updater app, so we have to be 100% that the new app replacement will be fully ready for 9.0.

**Comment by DeepDiver1975** [comment]
> Would it maybe be better to create a new app/repo for this ?

we did discuss this before - since the current updater app will stop to exist starting 9.0 this is the natural approach

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,26 +0,0 @@
-## Submitting issues
```
Comment (DeepDiver1975) [inline CONTRIBUTING.md]: we should keep this

**Comment by PVince81** [comment]
So people on web hosters (no CLI) will not be able to upgrade the code to 9.0 and will have to do it manually with the tarball ? And then with 9.1 or later there will be a new web UI so it will become easy again ?

(reading from https://github.com/owncloud/core/issues/18721#issuecomment-154418822)

> the first implementation will be console based only

**Comment by DeepDiver1975** [comment]
> So people on web hosters (no CLI) will not be able to upgrade the code to 9.0 and will have to do it manually with the tarball ? And then with 9.1 or later there will be a new web UI so it will become easy again ?

We need to have a web ui for the updater for 9.0 as well . we simply started out with the console.

**Comment by PVince81** [comment]
@VicDeo what's left to be done ?

**Comment by PVince81** [comment]
Would be good to have a list of checkboxes for all that it can already do and what's left to be done (for the 9.0 scoped things)

**Comment by VicDeo** [comment]
@PVince81 I guess no one made a decision what exactly is in the scope of 9.0
What I'm doing now is trying to gain functionality the current Updater app has.
I added the checkboxes according to all steps requested by @karlitschek. All of them have generic implementation atm. (and do nothing useful).

I  added checkpoint command, that is needed to revert unsuccessful upgrade attempt and some other things

**Comment by PVince81** [comment]
Okay thanks. So looks like there is still a lot to do.

In the meantime I guess that some could be called through `occ` first and then moved to this app later ?

**Comment by VicDeo** [comment]
@PVince81 according to @DeepDiver1975 proposal as much as possible (~ everything) should be implemented via `occ` command. 
e.g. `executeCoreUpgradeScripts` command just executes `occ upgrade`

**Comment by DeepDiver1975** [comment]
From a first run experience - there has to be an updater.sh or what so ever

**Comment by DeepDiver1975** [comment]
@karlitschek what do you think about adding the pure installation use case to this tool as well?

Installing an instance is basically the same as upgrade without the backup :wink:

**Comment by DeepDiver1975** [comment]
@VicDeo invocation of occ did fail for me because occ had no execution flag.
To be on the save side we might want to run php console.php instead ?

**Comment by DeepDiver1975** [comment]
we need to add a version check - I was testing this on a 8.0 instance and got:

````
$ ./application.php 


                                                                                                                        
  [UnexpectedValueException]                                                                                            
  Could not parse a response for config:list. Please check if the current shell user can run occ command. Raw output:   
  An unhandled exception has been thrown:                                                                               
  Exception: Not installed in /home/deepdiver/Development/ownCloud/test_updater/lib/base.php:223                        
  Stack trace:                                                                                                          
  #0 /home/deepdiver/Development/ownCloud/test_updater/lib/base.php(558): OC::checkInstalled()                          
  #1 /home/deepdiver/Development/ownCloud/test_updater/lib/base.php(1039): OC::init()                                   
  #2 /home/deepdiver/Development/ownCloud/test_updater/console.php(12): require_once('/home/deepdiver...')              
  #3 /home/deepdiver/Development/ownCloud/test_updater/occ(11): require_once('/home/deepdiver...')                      
  #4 {main}

**Comment by DeepDiver1975** [comment]
furthermore - we need to output anything more user friendly - no stack traces and clear sentences

**Comment by DeepDiver1975** [comment]
And this is where we require a version check :wink: 

````

[error] 

                                                                                                            
  [Symfony\Component\Process\Exception\ProcessFailedException]                                              
  The command "/home/deepdiver/Development/ownCloud/test_updater/occ config:list  --output "json"" failed.  
  Exit Code: 1(General error)                                                                               
  Output:                                                                                                   
  ================                                                                                          
  Error Output:                                                                                             
  ================                                                                                          
                                                                                                            
    [InvalidArgumentException]                                                                              
    There are no commands defined in the "config" namespace.                                                
                                                                                                            
                                                                                                            


Exception trace:
 () at /home/deepdiver/Development/ownCloud/test_updater/updater/src/Utils/OccRunner.php:47
 Owncloud\Updater\Utils\OccRunner->run() at /home/deepdiver/Development/ownCloud/test_updater/updater/src/Utils/OccRunner.php:53
 Owncloud\Updater\Utils\OccRunner->runJson() at /home/deepdiver/Development/ownCloud/test_updater/updater/src/Utils/ConfigReader.php:94
 Owncloud\Updater\Utils\ConfigReader->load() at /home/deepdiver/Development/ownCloud/test_updater/updater/src/Utils/ConfigReader.php:45
 Owncloud\Updater\Utils\ConfigReader->init() at /home/deepdiver/Development/ownCloud/test_updater/updater/src/Console/Application.php:92
 Owncloud\Updater\Console\Application->doRun() at /home/deepdiver/Development/ownCloud/test_updater/updater/vendor/symfony/console/Application.php:120
 Symfony\Component\Console\Application->run() at /home/deepdiver/Development/ownCloud/test_updater/updater/application.php:42

**Comment by DeepDiver1975** [comment]
- [ ] application.php list does not output the list of commands
- [ ] application.php without any argument should list an introduction and kick off the upgrade

**Comment by VicDeo** [comment]
@DeepDiver1975 it should, after the commit More stability if occ fails early (https://github.com/owncloud/updater/commit/23a4c2c5b639bda70791cff9cce61298f26fde2d)

**Comment by VicDeo** [comment]
I mean `application.php list` does not output the list of commands

**Comment by DeepDiver1975** [comment]
Let's merge this in now and continue with PRs against master to push the functionalities

**Pull request merged**